lib/puppet/http/service/file_server.rb

```diff
@@ -0,0 +1,103 @@
+require 'puppet/file_serving/metadata'
+
+class Puppet::HTTP::Service::FileServer < Puppet::HTTP::Service
+  API = '/puppet/v3'.freeze
+
+  EXCLUDED_FORMATS = [:yaml, :b64_zlib_yamli, :dot]
+
+  def initialize(client, server, port)
+    url = build_url(API, server || Puppet[:server], port || Puppet[:masterport])
+    super(client, url)
+  end
+
+  def get_file_metadata(mount_point:, path:, environment:,
+                        links: :manage, checksum_type: Puppet[:digest_algorithm], source_permissions: :ignore,
+                        ssl_context: nil)
+
+    model = Puppet::Transaction::Report
+    network_formats = model.supported_formats - EXCLUDED_FORMATS
+    mime_types = network_formats.map { |f| model.get_format(f).mime }
+
+    response = @client.get(
+      with_base_url("/file_metadata/#{mount_point}/#{path}"),
+      headers: add_puppet_headers({ 'ACCEPT' => mime_types.join(', ') }),
+      params: {
+        :links => links,
+        :checksum_type => checksum_type,
+        :source_permissions => source_permissions,
+        :environment => environment
+      },
+      ssl_context: ssl_context
+    )
+
+
+    if response.success?
+      header = response['Content-Type']
+      if header
+        header.gsub!(/\s*;.*$/,'') # strip any charset
+        formatter = Puppet::Network::FormatHandler.mime(header)
+        raise Puppet::HTTP::ProtocolError.new("Content-Type is unsupported") if EXCLUDED_FORMATS.include?(formatter.name)
+      else
+        raise Puppet::HTTP::ProtocolError.new(_("No content type in http response; cannot parse"))
+      end
+      Puppet::FileServing::Metadata.convert_from(formatter.name, response.body.to_s)
+    else
+      raise Puppet::HTTP::ResponseError.new(response)
+    end
+  end
+
+  def get_file_metadatas(mount_point:, path:, environment:,
+                        recurse: :false, ignore: nil,
+                        links: :manage, checksum_type: Puppet[:digest_algorithm], source_permissions: :ignore,
+                        ssl_context: nil)
+
+    model = Puppet::Transaction::Report
+    network_formats = model.supported_formats - EXCLUDED_FORMATS
+    mime_types = network_formats.map { |f| model.get_format(f).mime }
+
+    response = @client.get(
+      with_base_url("/file_metadatas/#{mount_point}/#{path}"),
+      headers: add_puppet_headers({ 'ACCEPT' => mime_types.join(', ') }),
+      params: {
+        :recurse => recurse,
+        :ignore => ignore,
```

Comment:
ignore is a multivalued query parameter, e.g. `ignore=CVS&ignore=.git&ignore=.hg`, so the default value should probaby be `[]`?

Reply:
Yesterday we set the default as `nil`, but I'm happy to change that

Reply:
Should we be looping through the given array and add an ignore header for each entry? If that's the case, defaulting to an empty array makes more sense

Reply:
I think `Net::HTTP` will auto-magically convert `:ignore => ['CVS', '.git']` to the stringified query parameters above, but it'd be good to test it

Reply:
I'm getting `ignore=%5B%22CVS%22,%20%22.git%22,%20%22.hg%22%5D` with `metadatas = subject.get_file_metadatas(mount_point: 'infinity', path: path, environment: environment, ignore: ['CVS', '.git', '.h    g'])`, so it doesn't look like it's going to be kind to us here

Reply:
I think there's some special logic in our current indirector/rest code to pass the headers to `Net::HTTP` as `['ignore', 'CVS', 'ignore', '.git']`, so we'll need to do something similar.